[max] nn: respect strict flag in Module.load_state_dict (fix PP staged loads)

dl-alexandre opened the pull request:
## Summary
- Fix `Module.load_state_dict` to respect `strict` for missing/unexpected keys.
- Previously raised on missing weights even when `strict=False`, breaking Pipeline Parallel staged loading.
- Now only raises when `strict=True`; otherwise tolerates missing/unexpected keys.

## Rationale
- Aligns with common frameworks (e.g., PyTorch) and unblocks PP.
- Verified with Llama3 PP where `rope_freqs.weight` or `lm_head.weight` may be absent per stage.

## Details
- Collect `missing_keys`; only raise if `strict=True`.
- Keep unexpected keys validation gated by `strict=True`.
- Updated docstring.

## Repro enabled
MAX_SERVE_OFFLINE_INFERENCE=1 \
CUDA_VISIBLE_DEVICES=0,1 \
max generate \
  --model-path meta-llama/Llama-3.2-1B-Instruct \
  --devices=gpu:0,1 \
  --pipeline-parallel-degree 2 \
  --max-length 8192 \
  --device-memory-utilization 0.6 \
  --max-batch-size 1 \
  --max-new-tokens 10 \
  --prompt "hi"

## Notes
- No API break for strict=True; strict=False becomes lenient as intended.
- Follow-up: add unit test exercising both strict modes.

Commit:
max.nn: respect strict flag in Module.load_state_dict for missing/unexpected keys

Pipeline-parallel loading failed because load_state_dict raised on missing
weights even when strict=False. This change:
- defers raising on missing keys unless strict=True
- keeps unexpected-keys validation under strict=True

This aligns behavior with PyTorch and fixes PP staged weight loading.

k-w-w commented on max/nn/layer/layer.py:
```diff
@@ -330,23 +334,34 @@ def load_state_dict(
+                            full_weight_name, state_dict.keys(), n=1
+                        ):
+                            msg += f" Did you mean '{possible_match[0]}'?"
+                        raise ValueError(msg)
```
Since an error is raised below, do we need to raise an error here?

k-w-w reviewed (commented):
Thanks for the PR! Agreed, missing weights should be caught only if strict=True.

dl-alexandre committed:
Update layer.py

Removed Immediate Error, Simplified Logic, Consolidated Error Reporting, and Improved "Did you mean?"

k-w-w reviewed (approved):
LGTM, thank you!
